subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -0,0 +1,622 @@
+
+  // Split the NALU into several FU
+  for (gsize offset = 0; offset < rbsp_size; offset += fu_pyl_max_size) {
+    gsize remaining = rbsp_size - offset;
+    gboolean last = remaining <= fu_pyl_max_size;
+    gsize size = last ? remaining : fu_pyl_max_size;
+
+    GstBuffer *fu_buf =
+        gst_rtp_h266_pay_extract_fu (rtph266pay, nalu, offset, size, last);
+    if (!fu_buf)
+      goto error;
+    GST_DEBUG_OBJECT (rtph266pay, "FU [%lu, %lu] %" GST_PTR_FORMAT,
+        offset, offset + size, fu_buf);
+    gst_buffer_list_add (out_buflist, fu_buf);
+  }
+
+  GST_DEBUG_OBJECT (rtph266pay, "Pushing FU GstBufferList %" GST_PTR_FORMAT,
+      out_buflist);
+  return gst_rtp_base_payload_push_list (rtpbasepay, out_buflist);
+error:
+  gst_buffer_list_unref (out_buflist);
+  return GST_FLOW_ERROR;
+}
+
+static gboolean
+gst_rtp_h266_pay_up_fits_in_mtu (GstRtpH266Pay * rtph266pay, Nalu * nalu)
+{
+  guint mtu = GST_RTP_BASE_PAYLOAD_MTU (rtph266pay);
+  guint payload_size = nalu->size;
+  // TODO: Consider DONL
+  gboolean fits = gst_rtp_buffer_calc_packet_len (payload_size, 0, 0) <= mtu;
+
+  if (!fits) {
+    GST_DEBUG_OBJECT (rtph266pay,
+        "NALU does not fit into %u MTU: %" GST_PTR_FORMAT, mtu, nalu->nalu_buf);
+  }
+  return fits;
+}
+
+static GstBuffer *
+gst_rtp_h266_pay_extract_fu (GstRtpH266Pay * rtph266pay, const Nalu * nalu,
+    gsize offset, gsize size, gboolean last)
+{
+  GstRTPBasePayload *rtpbasepay = GST_RTP_BASE_PAYLOAD (rtph266pay);
+  GstRTPBuffer rtp = GST_RTP_BUFFER_INIT;
+  gboolean au_end = last && nalu->au_end;
+  gboolean first = (offset == 0);
+  GstBuffer *fu_pyl_buf;
+  GstBuffer *fu_buf;
+  guint8 *fu_hdr;
+
+  fu_pyl_buf = gst_buffer_copy_region (nalu->rbsp_buf, GST_BUFFER_COPY_MEMORY,
+      offset, size);
+  if (!fu_pyl_buf)
+    return NULL;
+
+  // Allocate just the RTP header + PayloadHdr + FU header. We'll add the buffer
+  // payload later. This way we avoid unnecessary copies
+  fu_buf =
+      gst_rtp_base_payload_allocate_output_buffer (rtpbasepay, FU_HDR_LEN, 0,
+      0);
+  if (!gst_rtp_buffer_map (fu_buf, GST_MAP_WRITE, &rtp))
+    goto error_unref;
+
+  // Copy required buffer metadata
+  gst_buffer_copy_into (fu_buf, nalu->nalu_buf,
+      GST_BUFFER_COPY_FLAGS | GST_BUFFER_COPY_TIMESTAMPS, 0, -1);
+  gst_rtp_buffer_set_marker (&rtp, au_end);
+
+  fu_hdr = gst_rtp_buffer_get_payload (&rtp);
+  if (!fu_hdr)
+    goto error_unmap;
+
+  // Setup PayloadHdr and FU Header
+  // |     PayloadHdr (NALU HDR)     |   FU HEADER   |
+  // |-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-|-+-+-+-+-+-+-+-|
+  // |F|Z| LayerID   | Type(29)| TID |S|E|P|  FuType |
+  // +---------------+---------------|---------------+
+  guint8 read = gst_buffer_extract (nalu->hdr_buf, 0, fu_hdr, NALU_HDR_LEN);
+  if (read < NALU_HDR_LEN)
+    goto error_unmap;
+  fu_hdr[1] = (FU_TYPE << 3) | (fu_hdr[1] & 0x07);      // set Type = 29
+  fu_hdr[2] = (!!first << 7) | (!!last << 6) | (!!au_end << 5) | nalu->type;
```

Comment:
`(!!au_end << 5)` This au_end tells is the last NAL unit in an AU. 

However, according RFC9328, the P bit should be marked if it is the last **VCL** NAL unit:

```
   P: 1 bit
      When set to 1, the P bit indicates the last FU of the last VCL NAL
      unit of a coded picture, i.e., the last byte of the FU payload is
      also the last byte of the last VCL NAL unit of the coded picture.
      When the FU payload is not the last fragment of the last VCL NAL
      unit of a coded picture, the P bit MUST be set to 0.
```

So, if in an AU we had: `non-VCL | non-VCL | VCL | VCL | ... | VCL | non-VCL |`, and you are processing the last NALU (a non-VCL), according your code you would set the P bit to 1, but this should be set to 0.

Reply:
Honestly, I don't fully understand when I should set the `P` bit. We have to investigate this more deeply.

Reply:
I still do not have it fully clear, but after conversation https://fluendo.slack.com/archives/C2VDNCG69/p1731600383455059 I think it will be unlikely (or impossible) to find a non-VCL at the end of the AU. So this solution is good enough.